rust/agama-server/src/agama-web-server.rs

```diff
@@ -32,23 +60,213 @@ struct Cli {
+    }
+
+    // check that the key belongs to the certificate
+    tls_builder.check_private_key().unwrap();
+
+    tls_builder.build()
+}
+
+// build a response for the HTTP -> HTTPS redirection
+// returns 308 permanent redirect
+fn redirect_https(host: &str, uri: &hyper::Uri) -> Response<String> {
+    let builder = Response::builder()
+        // build the redirection target URL
+        .header("Location", format!("https://{}{}", host, uri))
+        .status(hyper::StatusCode::PERMANENT_REDIRECT);
+
+    builder.body(String::from("")).unwrap()
+}
+
+// build an error response for the HTTP -> HTTPS redirection when we cannot build
+// the redirect response from the original request
+// returns error 400
+fn redirect_error() -> Response<String> {
+    let builder = Response::builder().status(hyper::StatusCode::BAD_REQUEST);
+
+    let msg = "HTTP protocol is not allowed for external requests, please use HTTPS.\n";
+    builder.body(String::from(msg)).unwrap()
+}
+
+// build a router for the HTTP -> HTTPS redirection
+// if the redirection URL cannot be built from the original request it returns error 400
+// instead of the redirection
+fn https_redirect() -> Router {
+    // see https://docs.rs/axum/latest/axum/routing/struct.Router.html#example
+    let redirect_service = tower::service_fn(|req: axum::extract::Request| async move {
+        let request_host = req.headers().get("host");
+        match request_host {
+            // missing host in the request header, we cannot build the redirection URL, return error 400
+            None => Ok(redirect_error()),
+            Some(host) => {
+                let host_string = host.to_str();
+                match host_string {
+                    // invalid host value in the request
+                    Err(_) => Ok(redirect_error()),
+                    Ok(host_str) => return Ok(redirect_https(host_str, req.uri())),
+                }
+            }
+        }
+    });
+
+    Router::new()
+        // the wildcard path below does not match an empty path, we need to match it explicitly
+        .route_service("/", redirect_service)
+        .route_service("/*path", redirect_service)
+}
+
+// start the web server
+async fn start_server(address: String, service: Router, ssl_acceptor: SslAcceptor) {
+    tracing::info!("Starting Agama web server at {}", address);
+
+    // see https://github.com/tokio-rs/axum/blob/main/examples/low-level-openssl/src/main.rs
+    // how to use axum with openSSL
+    let listener = tokio::net::TcpListener::bind(&address)
+        .await
+        .unwrap_or_else(|_| panic!("could not listen on {}", &address));
+
+    pin_mut!(listener);
+
+    let redirector = https_redirect();
+
+    loop {
+        let tower_service = service.clone();
+        let redirector_service = redirector.clone();
+        let tls_acceptor = ssl_acceptor.clone();
+
+        // Wait for a new tcp connection
+        let (tcp_stream, addr) = listener.accept().await.unwrap();
```

Comment:
unwrap here also means immediate kill. Not sure if we can handle it more nicely.

Reply:
If the server cannot listen on the port then we cannot do much about it. :worried:
But normally this should not happen, when running from Live ISO the port should not be used by anything else and the server should have enough permissions.

This could possibly happen only during development where a crash is acceptable.

Reply:
You cannot do that much, but you can report an error. In this case it just crashes.